### PR TITLE
Dicom initial commit

### PR DESCRIPTION
**Pull request by marcelolerendegui**
@yongtang, much appreciated for the help on kicking this off!

### CODE REVIEW

**Comment by marcelolerendegui** [comment]
> I signed it!

**Comment by yongtang** [comment]
@marcelolerendegui  @Ouwen I think we are almost there. Can you update the PR to use the updated `dcmtk.BUILD` from the following (to fix a few macOS issues):

https://raw.githubusercontent.com/tensorflow/io/535b70ea5e23d5e8e4e9131fc700552f41cdea4d/third_party/dcmtk.BUILD

Once dcmtk.BUILD is updated I think we are ready to get this PR merged.

**Comment by yongtang** [comment]
@marcelolerendegui @Ouwen  There is another error:
```
E     File "/v/tests/test_dicom_eager.py", line 18
E       from __future__ import absolute_import
E   SyntaxError: from __future__ imports must occur at the beginning of the file
```

Can you also remove the empty line between the comment and the `from import`:
```
"""Tests for DICOM."""

from __future__ import absolute_import
```

**Comment by marcelolerendegui** [comment]
@yongtang Seems like nongnu.org is down.

File: tensorflow_io/dicom/kernels/decode_dicom_image.cc [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        frameWidth = image->getWidth();
+        frameHeight = image->getHeight();
+        pixelDepth = image->getDepth();
+        samples_per_pixel = image->isMonochrome() ? 1 : 3;
+
+        // Create an output tensor shape
+        TensorShape out_shape;
+        if ((samples_per_pixel == 1) && (color_dim == false))
+        {
+            out_shape = TensorShape({static_cast<int64>(frameCount), static_cast<int64>(frameHeight), static_cast<int64>(frameWidth)});
```
Comment (feihugis) [inline tensorflow_io/dicom/kernels/decode_dicom_image.cc]: The line length needs to be less than 80 chars. You can check the C++ code style using `clang-format`. More details can be found [here](https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#c-coding-style).

File: tensorflow_io/dicom/kernels/decode_dicom_data.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        DcmFileFormat dfile;
+        dfile.transferInit();
+        OFCondition cond = dfile.read(dataBuf);
+        dfile.transferEnd();
+
+        DcmDataset *dset = dfile.getDataset();
+        DcmMetaInfo *meta = dfile.getMetaInfo();
+
+        for (unsigned int tag_i = 0; tag_i < in_tags_flat.size(); tag_i++)
```
Comment (feihugis) [inline tensorflow_io/dicom/kernels/decode_dicom_data.cc]: `++tag_i` may be better.

File: tensorflow_io/dicom/kernels/decode_dicom_data.cc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            uint32 tag_value = in_tags_flat(tag_i);
+            uint16 tag_group_number = (uint16)((tag_value & 0xFFFF0000) >> 16);
+            uint16 tag_element_number = (uint16)((tag_value & 0x0000FFFF) >> 0);
+            DcmTag tag(tag_group_number, tag_element_number);
+
+            OFString val;
+            if (dset->tagExists(tag))
```
Comment (feihugis) [inline tensorflow_io/dicom/kernels/decode_dicom_data.cc]: The braces had better be added here

File: tensorflow_io/dicom/kernels/decode_dicom_image.cc [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    void uint64_to_t(uint64 in_value, unsigned int n_bits, double *out_value)
+    {
+        if (scale == "auto")
+            *out_value = (double)(in_value) / (double)((1ULL << n_bits) - 1);
+        else if (scale == "preserve")
+            *out_value = (double)(in_value);
+    }
+
+    string on_error;
```
Comment (feihugis) [inline tensorflow_io/dicom/kernels/decode_dicom_image.cc]: `on_error` -> `on_error_`.

File: tensorflow_io/dicom/kernels/decode_dicom_image.cc [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    string scale;
```
Comment (feihugis) [inline tensorflow_io/dicom/kernels/decode_dicom_image.cc]: ditto

File: tensorflow_io/dicom/kernels/decode_dicom_image.cc [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    bool color_dim;
```
Comment (feihugis) [inline tensorflow_io/dicom/kernels/decode_dicom_image.cc]: ditto

File: tensorflow_io/dicom/kernels/decode_dicom_image.cc [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+REGISTER_DECODE_DICOM_IMAGE_CPU(uint8);
+REGISTER_DECODE_DICOM_IMAGE_CPU(uint16);
+REGISTER_DECODE_DICOM_IMAGE_CPU(uint32);
+REGISTER_DECODE_DICOM_IMAGE_CPU(uint64);
+REGISTER_DECODE_DICOM_IMAGE_CPU(float);
+REGISTER_DECODE_DICOM_IMAGE_CPU(Eigen::half);
+REGISTER_DECODE_DICOM_IMAGE_CPU(double);
+
+#undef REGISTER_DECODE_DICOM_IMAGE_CPU
```
Comment (feihugis) [inline tensorflow_io/dicom/kernels/decode_dicom_image.cc]: missing new line

File: tensorflow_io/dicom/ops/decode_dicom_data_op.cc [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+using namespace tensorflow;
+
+REGISTER_OP("DecodeDICOMData")
+    .Input("contents: string")
+    .Input("tags: uint32")
+    .Output("tag_values: string")
+    .Doc(R"doc(
+loads a dicom file and returns the specified tags values as string.
+)doc");
```
Comment (feihugis) [inline tensorflow_io/dicom/ops/decode_dicom_data_op.cc]: missing newline

File: tensorflow_io/dicom/ops/decode_dicom_image_op.cc [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+REGISTER_OP("DecodeDICOMImage")
+    .Input("contents: string")
+    .Output("output: dtype")
+    .Attr("dtype: {uint8, uint16, uint32, uint64, float16, float, double} = DT_UINT16")
+    .Attr("color_dim: bool = true")
+    .Attr("on_error: {'strict', 'skip', 'lossy'} = 'skip'")
+    .Attr("scale: {'auto', 'preserve'} = 'preserve'")
+    .Doc(R"doc(
+loads a dicom image file and returns its pixel information in the specified output format
+)doc");
```
Comment (feihugis) [inline tensorflow_io/dicom/ops/decode_dicom_image_op.cc]: missing newline

File: tensorflow_io/dicom/python/ops/dicom_ops.py [hunk 10]
```diff
@@ -0,0 +1,4076 @@
...
+  ConcatenationFrameOffsetNumber = int('00209228', 16)
+  FunctionalGroupPrivateCreator = int('00209238', 16)
+  NominalPercentageofCardiacPhase = int('00209241', 16)
+  NominalPercentageofRespiratoryPhase = int('00209245', 16)
+  StartingRespiratoryAmplitude = int('00209246', 16)
+  StartingRespiratoryPhase = int('00209247', 16)
+  EndingRespiratoryAmplitude = int('00209248', 16)
+  EndingRespiratoryPhase = int('00209249', 16)
+  RespiratoryTriggerType = int('00209250', 16)
+  RRIntervalTimeNo
```
Comment (feihugis) [inline tensorflow_io/dicom/python/ops/dicom_ops.py]: missing newline

File: third_party/dcmtk.BUILD [hunk 11]
```diff
@@ -0,0 +1,990 @@
...
+        "#define DCMTK_FLOAT_TINYNESS_BEFORE OFTrue",
+        "#define DCMTK_DOUBLE_TINYNESS_BEFORE OFTrue",
+        "#define DCMTK_FLOAT_HAS_DENORM_LOSS OFTrue",
+        "#define DCMTK_DOUBLE_HAS_DENORM_LOSS OFTrue",
+        "#define DCMTK_ROUND_STYLE 1",
+        "",
+        "#endif // CONFIG_ARITH_H",
+        "EOF",
+    ]),
+)
```
Comment (feihugis) [inline third_party/dcmtk.BUILD]: missing newline

**Comment by yongtang** [comment]
All review comments have been addressed and all tests passed, except for the additional lint issue (one with python and another one with bazel):
```
************* Module tests.test_dicom_eager
W: 25, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 29, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 31, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 32, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 33, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
W: 37, 0: Bad indentation. Found 4 spaces, expected 2 (bad-indentation)
```

```
--- ./tensorflow_io/dicom/BUILD	2019-07-18 21:52:16.879995001 +0000
+++ /tmp/buildifier-tmp-258320356	2019-07-18 21:52:56.607995001 +0000
@@ -25,4 +25,4 @@
         "@local_config_tf//:libtensorflow_framework",
         "@local_config_tf//:tf_header_lib",
     ],
-)
\ No newline at end of file
+)
exit status 1
```

@marcelolerendegui can you update the PR again?

**Comment by yongtang** [comment]
Never mind, I will merge and add a follow up PR for the lint fix. @marcelolerendegui Thanks for the contribution 👍 🎉 🎆 !

**Comment by marcelolerendegui** [comment]
Hey @yongtang 

I added sample dicom files and tests.

I found a problem when running the tests: "Warning: no data dictionary loaded"
Aparently DCMTK is looking for the data dictionaries in $DCMDICTPATH.
I changed dcmtk.BUILD to define ENABLE_BUILTIN_DICTIONARY and now it is working.

I downloaded the 21 DCM sample files (totaling 8.3MB) from [here](https://barre.dev/medical/samples/)

Before requesting a pull:
Can these sample files be merged into tensorflow/io?

**Comment by yongtang** [comment]
@marcelolerendegui 8.3 MB is a little larger than I expected. Would prefer to download during the test than save it to git which will be in the history.

But if the files are not available through direct download then it should be fine.

**Pull request merged**